packages/msw-addon/src/mswDecorator.ts

```diff
@@ -0,0 +1,99 @@
+import { isNodeProcess } from 'is-node-process'
+import type { DecoratorFunction, StoryContext } from '@storybook/addons'
+import type { SetupWorkerApi, RequestHandler } from 'msw'
+import type { SetupServerApi } from 'msw/node'
+
+export type SetupApi = SetupWorkerApi | SetupServerApi
+export type InitializeOptions =
+  | Parameters<SetupWorkerApi['start']>[0]
+  | Parameters<SetupServerApi['listen']>[0]
+
+export type DecoratorParameters = {
+  msw:
+    | RequestHandler[]
+    | { handlers: RequestHandler[] | Record<string, RequestHandler> }
+}
+
+export type DecoratorContext = StoryContext & {
+  parameters: DecoratorParameters
+}
+
+const IS_BROWSER = !isNodeProcess()
+let api: SetupApi
+
+export function initialize(options?: InitializeOptions): SetupApi {
+  if (IS_BROWSER) {
+    const { setupWorker } = require('msw')
+    const worker = setupWorker()
+    worker.start(options)
+    api = worker
+  } else {
+    /**
+     * Webpack 5 does not provide node polyfills as it did before.
+     * Also, it can't tell whether a code will be executed at runtime, so it has to process everything.
+     * This branch of the conditional statement will NEVER run in the browser, but Webpack can't know so
+     * it breaks builds unless we start providing node polyfills.
+     *
+     * As a workaround, we use __non_webpack_require__ to tell Webpack to ignore this, and we define it
+     * to globalThis so it works correctly when running in node.
+     * @see https://github.com/webpack/webpack/issues/8826#issuecomment-660594260
+     */
+    globalThis.__non_webpack_require__ = require
+
+    const { setupServer } = __non_webpack_require__('msw/node')
+    const server = setupServer()
+    server.listen(options)
+    api = server
+  }
+
+  return api
+}
+
+export function initializeWorker(options?: InitializeOptions): SetupApi {
+  console.warn(
+    `[MSW] "initializeWorker" is now deprecated, please use "initialize" instead. This method will be removed in future releases.`
+  )
+  return initialize(options)
+}
+
+export function getWorker(): SetupApi {
+  if (api === undefined) {
+    throw new Error(
+      `[MSW] Failed to retrieve the worker: no active worker found. Did you forget to call "initialize"?`
+    )
+  }
+
+  return api
+}
+
+export const mswDecorator: DecoratorFunction = (
+  storyFn,
+  context: DecoratorContext
+) => {
+  const {
+    parameters: { msw },
+  } = context
+
+  if (api) {
+    api.resetHandlers()
+
+    if (msw) {
+      if (Array.isArray(msw) && msw.length > 0) {
+        // Support an Array of request handlers (backwards compatability).
+        api.use(...msw)
+      } else if ('handlers' in msw) {
```

Comment:
We've had the `'handlers' in msw && msw.handlers` here previously. Does that mean we expect `msw.handlers` to be null, potentially? If we do, let's reflect that in the type definitions.

Reply:
This is mostly for Typescript safety actually, so the type inference works correctly. I don't see a scenario where people would use msw and not pass handlers (only the deprecated one), but these types are for internal purposes only. There is currently no proper way of handling addon parameter's types in Storybook unfortunately. People can do some stuff with the `DecoratorParameters` type, so it's nice that we export them for future-sake

Reply:
Using `in` provides the correct type inference here.

Reply:
We can add a check if handlers is null or undefined! I think I found a possible scenario:

1 - User defines handlers globally:
```
export parameters = {
  msw: {
    handlers: [ ...someHandlers ]
  }
}
```

Then, in certain story, they want to disable:
```
Story.parameters = { msw: { handlers: null } }
```

Even though that's not the recommended way of handling this situation, users might end up doing that and the code would break in ` Object.values(msw.handlers) ` so it's good to check that!

Reply:
Fixed!